service/tablet_allocator.cc

```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+        // means that scaling may not be effective and in the worst case we may overshoot the goal by
+        // a factor of 2. This is acceptable since the goal is a soft limit and not a hard constraint.
+        // Scaling post-alignment would be problematic. If we scale down all tables fairly, we undershoot the goal
+        // by a factor of 2 in the worst case. If we choose a subset of tables to scale down by a factor of 2 then
+        // we have a problem of making sure that the choice is stable across scheduler invocations to avoid
+        // oscillations of decisions.
+
+        std::unordered_map<table_id, double> table_scaling;
+
+        std::unordered_map<sstring, unsigned> shards_per_dc;
+        _tm->get_topology().for_each_node([&] (const node& n) {
+            if (n.is_normal() && _tm->is_normal_token_owner(n.host_id())) {
+                shards_per_dc[n.dc_rack().dc] += n.get_shard_count();
+            }
+        });
+
+        for (auto&& [dc, shard_count] : shards_per_dc) {
+            double old_avg_tablets_per_shard = 0;
+            double avg_tablets_per_shard = 0;
+
+            for (auto&& [table, table_plan] : plan.tables) {
+                auto* rs = rs_by_table[table];
+                if (!rs) {
+                    continue;
+                }
+
+                auto rf = rs->get_replication_factor(dc);
+                auto get_avg_tablets_per_shard = [&] (size_t tablet_count) {
+                    return double(tablet_count) * rf / shard_count;
+                };
+
+                auto old_tablets_per_shard = get_avg_tablets_per_shard(table_plan.current_tablet_count);
+                old_avg_tablets_per_shard += old_tablets_per_shard;
+                lblogger.debug("old_avg_tablets_per_shard [dc={}, table={}]: {:.3f}", dc, table, old_tablets_per_shard);
+
+                auto tablets_per_shard = get_avg_tablets_per_shard(table_plan.target_tablet_count);
+                avg_tablets_per_shard += tablets_per_shard;
+                lblogger.debug("new_avg_tablets_per_shard [dc={}, table={}]: {:.3f}", dc, table, tablets_per_shard);
+            }
+
+            if (!new_table) {
+                bool overloaded = old_avg_tablets_per_shard > _tablets_per_shard_goal;
+                lblogger.info("avg_tablets_per_shard[dc={}]: {:.3f}{}", dc, old_avg_tablets_per_shard,
+                    overloaded ? " (overloaded!)" : "");
+            }
+
+            bool overloaded = avg_tablets_per_shard > _tablets_per_shard_goal;
+            lblogger.debug("new_avg_tablets_per_shard[dc={}]: {:.3f}{}", dc, avg_tablets_per_shard,
+                overloaded ? " (overloaded!)" : "");
+
+            if (overloaded) {
+                auto scale = _tablets_per_shard_goal / avg_tablets_per_shard;
+
+                for (auto&& [table, table_plan]: plan.tables) {
+                    auto *rs = rs_by_table[table];
+                    if (!rs) {
```

Comment:
again, how can that happen?

Reply:
Just in test_mode?

Reply:
Yes, in test_mode (which is going away).